src/clients/c++/perf_client/request_rate_manager.cc

```diff
@@ -26,6 +26,24 @@
 
 #include "src/clients/c++/perf_client/request_rate_manager.h"
 
+RequestRateManager::~RequestRateManager()
+{
+  early_exit = true;
+  // wake up all threads
+  wake_signal_.notify_all();
+
+  size_t cnt = 0;
+  for (auto& thread : threads_) {
+    thread.join();
+    if (!threads_stat_[cnt]->status_.IsOk()) {
+      std::cerr << "Thread [" << cnt
+                << "] had error: " << (threads_stat_[cnt]->status_)
+                << std::endl;
+    }
+    cnt++;
+  }
+}
```

Comment:
The code in both derived class destructors are the same, why move it out of the base class destructor?

Reply:
This is done because of the order in which destructors are called. The base class destructor will be called after the derived class. The worker threads depends on certain data members of derived class to operate and exit gracefully. So, the destruction of derived class should wait for the threads to finish before exiting. I don't want entire code to be replicated in all the derived classes hence, I moved the final clean-up to the base class.

Reply:
Can you add inline comment about this?

Reply:
Or better can you create a "StopWorkerThreads" or similar utility function in the base class that has this code and then call it from both derived destructors.